nncf/openvino/quantization/cache_statistics.py

```diff
@@ -0,0 +1,169 @@
+# Copyright (c) 2024 Intel Corporation
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#      http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Dict, List, Optional, Tuple
+
+import openvino.runtime as ov
+
+from nncf.common.factory import NNCFGraphFactory
+from nncf.common.factory import StatisticsAggregatorFactory
+from nncf.common.graph.graph import NNCFGraph
+from nncf.common.graph.graph import NNCFNode
+from nncf.common.tensor_statistics.aggregator import StatisticsAggregator
+from nncf.data import Dataset
+from nncf.openvino.graph.model_utils import remove_friendly_name_duplicates
+from nncf.parameters import BackupMode
+from nncf.parameters import CompressWeightsMode
+from nncf.parameters import SensitivityMetric
+from nncf.quantization.advanced_parameters import AdvancedCompressionParameters
+from nncf.quantization.algorithms.weight_compression.algorithm import WeightCompression
+from nncf.quantization.algorithms.weight_compression.gptq import GPTQ
+from nncf.quantization.algorithms.weight_compression.mixed_precision import MIXED_PRECISION_CRITERIA
+from nncf.scopes import IgnoredScope
+
+
+def register_statistics_for_algorithm(
```

Comment:
> Please avoid using "private" methods from the compression_algorithm. At least, I would suggest to remove "_" for such methods.

done

> You implemented logic of collection statistics for caching using loop with creating algorithm with different parameters and calling get_statistic_points. I guess this is sub-optimal implementation because it creates additional loops by the target model or target nodes.

I followed the `Algorithm` interface declaration to keep the implementation clear and aligned with the existing structure. The current non-optimality arises because each call to `get_statistic_points()` for every `MixedPrecision` algorithm results in repeated work. However, this method already reuses the `matmul_input_to_output_nodes_map`, and the inefficiency comes mainly from the repeated calls (4 times instead of 1).

Regarding this specific piece of code:
```
n_dims = len(graph.get_output_edges_by_port_id(act_node, output_port_id)[0].tensor_shape)
if n_dims < 2:
    raise RuntimeError(
        f"Data-aware mixed precision criteria are not supported for MatMuls with 1D inputs. "
        f"Node: {act_node.node_name}, number of dimensions: {n_dims}."
    )
```
https://github.com/openvinotoolkit/nncf/pull/3017/files#diff-130add17d0ec2c31bfb50f061cd45c31006a8b225391254967dedca9752b7a67R240-R245

I don’t believe this is a major issue, as it only introduces a minor overhead. Optimizing it might not provide significant performance gains, given the context. What are your thoughts on this?